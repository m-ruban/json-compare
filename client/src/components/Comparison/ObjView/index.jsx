import React from 'react';
import PropTypes from 'prop-types';
import { Col } from 'reactstrap';
import Prop from 'components/JSON/Prop';

const ObjView = ({ obj, showAlerts, showRequired }) => (
    <Col xs="7">
        <div className="json result without-panel">
            <Prop value={obj} showAlerts={showAlerts} showRequired={showRequired} lvl={1} />
        </div>
    </Col>
);

ObjView.propTypes = {
    obj: PropTypes.object,
    showAlerts: PropTypes.bool,
    showRequired: PropTypes.bool,
};

ObjView.defaultProps = {
    showAlerts: false,
    showRequired: false,
};

export default ObjView;
