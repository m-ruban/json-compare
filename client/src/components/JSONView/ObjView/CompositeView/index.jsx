import React, { Fragment, useCallback, useState } from 'react';
import PropTypes from 'prop-types';

import { BRACKET_ARRAY_CLOSE, BRACKET_ARRAY_OPEN, BRACKET_OBJECT_CLOSE, BRACKET_OBJECT_OPEN } from 'constants/compare';
import Icon from 'components/JSONView/ObjView/CompositeView/Icon';
import Value from 'components/JSONView/ObjView/CompositeView/Value';

const ObjView = ({ obj, lvl, path, showAlerts, showRequired }) => {
    const isArray = Array.isArray(obj);
    const [isOpen, setOpen] = useState(true);
    const toggle = useCallback(() => {
        setOpen(!isOpen);
    }, [isOpen]);

    return (
        <Fragment>
            <Icon isOpen={isOpen} toggle={toggle} />
            {isArray ? BRACKET_ARRAY_OPEN : BRACKET_OBJECT_OPEN}
            {isOpen ? (
                <Value
                    isArray={isArray}
                    obj={obj}
                    path={path}
                    lvl={lvl}
                    showAlerts={showAlerts}
                    showRequired={showRequired}
                />
            ) : null}
            {isArray ? BRACKET_ARRAY_CLOSE : BRACKET_OBJECT_CLOSE}
        </Fragment>
    );
};

ObjView.propTypes = {
    obj: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
    lvl: PropTypes.number,
    path: PropTypes.array,
    showAlerts: PropTypes.bool,
    showRequired: PropTypes.bool,
};

export default ObjView;
