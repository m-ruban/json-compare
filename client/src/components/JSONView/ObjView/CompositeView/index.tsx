import React from 'react';

import Value from 'components/JSONView/ObjView/CompositeView/Value';
import Mark from 'components/Mark';
import {
    BRACKET_ARRAY_CLOSE,
    BRACKET_ARRAY_OPEN,
    BRACKET_OBJECT_CLOSE,
    BRACKET_OBJECT_OPEN,
    CLOSE_ICON,
    OPEN_ICON,
} from 'constants/compare';

interface ICompositeViewProps {
    obj: any[] | object;
    lvl: number;
    path: string[];
    showAlerts: boolean;
    showRequired: boolean;
}

const CompositeView = ({ obj, lvl, path, showAlerts, showRequired }: ICompositeViewProps) => {
    const isArray = Array.isArray(obj);
    const [isOpen, setOpen] = React.useState(true);
    const toggle = React.useCallback(() => {
        setOpen(!isOpen);
    }, [isOpen]);

    return (
        <>
            <Mark type="switch" text={isOpen ? OPEN_ICON : CLOSE_ICON} onClick={toggle} />
            {isArray ? BRACKET_ARRAY_OPEN : BRACKET_OBJECT_OPEN}
            {isOpen ? (
                <Value
                    isArray={isArray}
                    obj={obj}
                    path={path}
                    lvl={lvl}
                    showAlerts={showAlerts}
                    showRequired={showRequired}
                />
            ) : null}
            {isArray ? BRACKET_ARRAY_CLOSE : BRACKET_OBJECT_CLOSE}
        </>
    );
};

export default CompositeView;
