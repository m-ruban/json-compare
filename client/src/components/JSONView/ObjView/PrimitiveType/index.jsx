import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import withConsumer from 'hoc/withConsumer';
import { STRING_TYPE } from 'constants/types';
import Alert from 'components/JSONView/ObjView/PrimitiveType/Alert';

const PrimitiveType = ({ value, showAlerts, showRequired, path, alerts }) => {
    const mark = typeof value === STRING_TYPE ? '"' : null;
    const key = path.join('-');
    const rValue = value !== null ? value.toString() : 'null';
    return (
        <Fragment>
            {mark}
            {rValue}
            {mark}
            <Alert id={key} alert={alerts.get(key)} showAlerts={showAlerts} showRequired={showRequired} />
        </Fragment>
    );
};

PrimitiveType.propTypes = {
    value: PropTypes.oneOfType([PropTypes.bool, PropTypes.string, PropTypes.number]),
    path: PropTypes.array,
    showAlerts: PropTypes.bool,
    showRequired: PropTypes.bool,
    alerts: PropTypes.instanceOf(Map),
};

export default withConsumer(PrimitiveType);
