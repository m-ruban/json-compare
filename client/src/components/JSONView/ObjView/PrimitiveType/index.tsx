import React, { Fragment } from 'react';

import Alert from 'components/JSONView/ObjView/PrimitiveType/Alert';
import { IAlertMap } from 'constants/interfaces';
import { STRING_TYPE } from 'constants/types';
import withConsumer from 'hoc/withConsumer';

interface IPrimitiveTypeProps extends IAlertMap {
    value: boolean | string | number;
    showAlerts: boolean;
    showRequired: boolean;
    path: string[];
}

const PrimitiveType = ({ value, showAlerts, showRequired, path, alerts }: IPrimitiveTypeProps) => {
    const mark = typeof value === STRING_TYPE ? '"' : null;
    const key = path.join('-');
    const rValue = value !== null ? value.toString() : 'null';
    return (
        <span>
            {mark}
            {rValue}
            {mark}
            <Alert alert={alerts.get(key)} showAlerts={showAlerts} showRequired={showRequired} />
        </span>
    );
};

export default withConsumer(PrimitiveType);
