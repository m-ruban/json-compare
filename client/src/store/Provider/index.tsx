import * as React from 'react';

import { COMPARE_REQ } from 'constants/compare';
import { EQUALITY_TYPE_STRICT } from 'constants/equality';
import { AUTHOR_LINK, FEEDBACK_LINK } from 'constants/links';
import { LOG_INFO } from 'constants/log';
import { Provider } from 'store/Context';

import { equalAlerts } from 'modules/compare';
import { parseAndLog } from 'modules/parse';

interface IProviderProps {
    children: any;
}

interface IProviderState {
    alerts: Map<string, object>;
    author: string;
    changeEqualityTypes: (event: React.SyntheticEvent) => void;
    changeObj: (event: React.SyntheticEvent) => void;
    changeTool: (value: string) => void;
    equalityTypes: string;
    feedback: string;
    isCompare: boolean;
    log: object;
    objObject: object;
    objStr: string;
    toggleCompare: (event: React.SyntheticEvent) => void;
    toolObject: object;
    toolStr: string;
}

class AppProvider extends React.Component<IProviderProps, IProviderState> {
    constructor(props) {
        super(props);
        this.state = {
            alerts: new Map(),
            author: AUTHOR_LINK,
            changeEqualityTypes: this.changeEqualityTypes.bind(this),
            changeObj: this.changeObj.bind(this),
            changeTool: this.changeTool.bind(this),
            equalityTypes: EQUALITY_TYPE_STRICT,
            feedback: FEEDBACK_LINK,
            isCompare: false,
            log: { text: '', type: null },
            objObject: {},
            objStr: '',
            toggleCompare: this.toggleCompare.bind(this),
            toolObject: {},
            toolStr: '',
        };
    }

    public toggleCompare() {
        const log = {
            text: 'toggle compare',
            type: LOG_INFO,
        };

        this.setState({ isCompare: !this.state.isCompare, log }, this.changeAlerts);
    }

    public changeEqualityTypes(event) {
        const { value } = event.target;
        const log = {
            text: 'change equality types',
            type: LOG_INFO,
        };
        this.setState({ equalityTypes: value, log }, this.changeAlerts);
    }

    public changeTool(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ toolObject: obj, toolStr: value, log }, this.changeAlerts);
    }

    public changeObj(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ objObject: obj, objStr: value, log }, this.changeAlerts);
    }

    public changeAlerts() {
        const { toolObject, objObject, equalityTypes, isCompare } = this.state;
        const isEmpty = (obj) => Object.keys(obj).length === 0;

        if (!isEmpty(toolObject) && !isEmpty(objObject) && isCompare) {
            const alerts = equalAlerts(toolObject, objObject, equalityTypes);
            let req = 0;
            let diff = 0;
            let text = '';

            alerts.forEach((alert) => {
                if (!alert.scalar) { return; }
                if (alert.res === COMPARE_REQ) {
                    req++;
                } else {
                    diff++;
                }
            });

            if (diff > 0 || req > 0) {
                text = `${diff} alert${diff > 1 ? 's' : ''}, ${req} required`;
            } else {
                text = 'objects are equal';
            }

            this.setState({
                alerts,
                log: { text, type: LOG_INFO },
            });
        }
    }

    public render() {
        return <Provider value={this.state}>{this.props.children}</Provider>;
    }
}

export default AppProvider;
