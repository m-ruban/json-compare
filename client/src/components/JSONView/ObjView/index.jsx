import React from 'react';
import PropTypes from 'prop-types';
import PrimitiveType from 'components/JSONView/ObjView/PrimitiveType';
import CompositeView from 'components/JSONView/ObjView/CompositeView';

const Property = ({ value, lvl, showAlerts, showRequired, path }) => {
    if (typeof value === 'object' && value) {
        return (
            <CompositeView obj={value} path={path} lvl={lvl + 1} showAlerts={showAlerts} showRequired={showRequired} />
        );
    }
    return <PrimitiveType path={path} value={value} showAlerts={showAlerts} showRequired={showRequired} />;
};

Property.propTypes = {
    value: PropTypes.any,
    lvl: PropTypes.number,
    path: PropTypes.array,
    showAlerts: PropTypes.bool,
    showRequired: PropTypes.bool,
};

Property.defaultProps = {
    lvl: 1,
    path: [],
};

export default Property;
