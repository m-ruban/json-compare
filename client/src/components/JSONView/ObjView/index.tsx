import React from 'react';

import CompositeView from 'components/JSONView/ObjView/CompositeView';
import PrimitiveType from 'components/JSONView/ObjView/PrimitiveType';

interface IObjViewProps {
    value: any[] | object;
    lvl: number;
    path: string[];
    showAlerts: boolean;
    showRequired: boolean;
}

const ObjView = ({ value, lvl, showAlerts, showRequired, path }: IObjViewProps) => {
    if (typeof value === 'object' && value) {
        return (
            <CompositeView obj={value} path={path} lvl={lvl + 1} showAlerts={showAlerts} showRequired={showRequired} />
        );
    }
    return <PrimitiveType path={path} value={value} showAlerts={showAlerts} showRequired={showRequired} />;
};

ObjView.defaultProps = {
    lvl: 1,
    path: [],
};

export default ObjView;
