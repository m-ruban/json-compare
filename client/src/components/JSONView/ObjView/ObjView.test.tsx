import React from 'react';

import ObjView from 'components/JSONView/ObjView';
import { render } from 'enzyme';
import { Provider } from 'store/Context';
import alerts from 'tests/mocks/alerts';

describe('<ObjView />', () => {
    it('Renders correctly', () => {
        const props = { value: [], showAlerts: false, showRequired: false, path: [] };
        const output = render(
            <Provider value={alerts}>
                <ObjView {...props} />
            </Provider>,
        );
        expect(output).toMatchSnapshot();
    });
    it('Renders correctly (obj)', () => {
        const props = { value: {id: 1}, showAlerts: false, showRequired: false, path: [] };
        const output = render(
            <Provider value={alerts}>
                <ObjView {...props} />
            </Provider>,
        );
        expect(output).toMatchSnapshot();
    });
});
