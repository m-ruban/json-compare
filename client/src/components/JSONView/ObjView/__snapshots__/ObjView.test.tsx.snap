// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ObjView /> Renders correctly (obj) 1`] = `
Array [
  <span
    class="json-mark json-mark_switch badge badge-json"
  >
    -
  </span>,
  "{",
  <div
    style="margin-left:12px"
  >
    <span>
      "id"
    </span>
    :Â 
    <span
      class="json-mark json-mark_number badge badge-json"
    >
      number
    </span>
    <span>
      1
    </span>
  </div>,
  "}",
]
`;

exports[`<ObjView /> Renders correctly 1`] = `
Array [
  <span
    class="json-mark json-mark_switch badge badge-json"
  >
    -
  </span>,
  "{]",
]
`;
